### PR TITLE
Optimize measurement integrator and ROS node performance

### PR DESCRIPTION
**Pull request by victorreijgwart**
# Description

This PR makes wavemap 20% faster for LiDAR inputs and adds support for [Tracy Profiler](https://github.com/wolfpld/tracy).

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### New features
* Annotate code for profiling with Tracy Profiler
* Switch to custom atan2 in LiDAR projection models
  * Speeds up wavemap by 20% overall for LiDAR inputs
  * No compromise in accuracy (slightly improves AUC, accuracy and recall)
* Minor general optimizations
* Add option to enable DCHECKs even when not compiling in debug mode
* Improve error messages when reading/writing a file fails

### Notes on Tracy
Being a hybrid profiler, Tracy combines the strengths of frame and sampling-based profiling. It also runs cross-platform, making it very useful as a strong general starting point for performance optimization. These stats can be complemented by tools like Intel's VTune or Nvidia's NSight when more detailed insights for specific platforms are desired.

To simplify the use of Tracy with ROS, we provide a wrapper that makes its library available as a [catkin package](https://github.com/ethz-asl/tracy_catkin). The application being profiled must link against this library. The package also includes scripts to easily build and run Tracy's GUI in Docker.

By default, Tracy is disabled and introduces zero overhead. To enable profiling with Tracy, add `-DTRACY_ENABLE=ON` to catkin's cmake-args and rebuild wavemap_all. To disable Tracy later, change the cmake argument to `-DTRACY_ENABLE=OFF`. It is also possible to remove the argument, but then you have to clean the catkin workspace (`catkin clean -y`) to make sure the argument doesn't stay in CMake's caches.

### Necessary upgrade steps
For catkin users, the [tracy_catkin](https://github.com/ethz-asl/tracy_catkin) package should be added to catkin_ws/src, either with
* `git clone git@github.com:ethz-asl/tracy_catkin.git`       # When using SSH [recommended]
* `git clone https://github.com/ethz-asl/tracy_catkin.git`  # For HTTPS

The new code can then be pulled and built as usual, with `catkin build wavemap_all`.

For Docker users, the only step is to pull the new image once the PR is merged and the CD pipeline publishes the new release. The image will already contain the tracy_catkin dependency.

# Testing

Unit tests are present to ensure that the approximate atan2's worst-case error is always within 0.0001 degrees of std::atan2.

We manually validated that this PR improves performance without affecting accuracy. The accuracy actually slightly improves.
Since we already have unit tests that ensure that all integration methods and data structures yield equivalent results, we only manually evaluate the change for the hashed chunked wavelet octree data structure and integrator.

|      Atan2       | ROC AUC | RAM usage | Wall time |
| -------------- | ----------- | ------------ | ---------- |
|   std::atan2    |     0.968    |     797.79     |     5.42     |
| approximate |     0.970    |     798.34     |     4.19     |

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any required changes in dependencies have been committed and pushed

### CODE REVIEW

File: libraries/wavemap_io/include/wavemap_io/file_conversions.h
Comment (LionelOtt) [inline libraries/wavemap_io/include/wavemap_io/file_conversions.h:1]: Technically could have used std::filesystem::path for the paths, though that's likely overkill here, given you don't do any of the path mangling that the class allows.

**Review by LionelOtt — APPROVED** [review]
Looks good to me. A general question was about Tracy, is that always active or is it compile time turned on/off?

File: libraries/wavemap_io/include/wavemap_io/file_conversions.h
Comment (victorreijgwart) [reply libraries/wavemap_io/include/wavemap_io/file_conversions.h:1]: I'll look into it. One advantage could also be that it better signals the intent than `std::string` does.

**Comment by victorreijgwart** [comment]
> Looks good to me. A general question was about Tracy, is that always active or is it compile time turned on/off?

Tracy has to be enabled with a compile time CMake flag and is off by default, in which case it fully disappears.

**Review by LionelOtt — APPROVED** [review]
Changes look good to me, no further comments.

**Pull request merged**